### PR TITLE
Apple pay completion callback

### PR DESCRIPTION
**Pull request by d4rkd3v1l**
**Description**
Currently one does not get notified, when the Apple Pay component (Apple Pay overlay) gets canceled/dismissed, e.g. when the user taps the cancel button or dismisses the overlay but tapping the empty space above the overlay.
I'm not sure if this is the preferred solution. I would be pretty happy if we'll see this or a similar solution in the main repo/project in the near future.

**Tested scenarios**
Tapping the cancel button as well as dismissing the overlay by tapping on the background leads to execution of the added completion callback.

**Fixed issue**: ?

### CODE REVIEW

File: Adyen/Components/Apple Pay/ApplePayComponent.swift [hunk 1]
```diff
@@ -96,7 +96,8 @@ public class ApplePayComponent: NSObject, PaymentComponent, PresentableComponent
                 merchantIdentifier: String,
                 summaryItems: [PKPaymentSummaryItem],
                 requiredBillingContactFields: Set<PKContactField> = [],
-                requiredShippingContactFields: Set<PKContactField> = []) throws {
+                requiredShippingContactFields: Set<PKContactField> = [],
+                completion: (() -> Void)? = nil) throws {
```
[the remark this answers is not in the document]
Comment (d4rkd3v1l) [reply Adyen/Components/Apple Pay/ApplePayComponent.swift:101]: Hey there, thanks for the review. I added documentation for those new parameters, but kept the naming as it already is being used like this for `stopLoading`. The internal property is also called `dismissCompletion`, so for me it kinda fits very well like this.

File: Adyen/Components/Apple Pay/ApplePayComponent.swift [hunk 2]
```diff
@@ -137,7 +139,7 @@ public class ApplePayComponent: NSObject, PaymentComponent, PresentableComponent
     /// - Parameter merchantIdentifier: The merchant identifier.
     /// - Parameter summaryItems: The line items for this payment.
     @available(*, deprecated, message: "Use init(paymentMethod:payment:merchantIdentifier:summaryItems:) instead.")
-    public init?(paymentMethod: ApplePayPaymentMethod, merchantIdentifier: String, summaryItems: [PKPaymentSummaryItem]) {
+    public init?(paymentMethod: ApplePayPaymentMethod, merchantIdentifier: String, summaryItems: [PKPaymentSummaryItem], completion: (() -> Void)? = nil) {
```
[the remark this answers is not in the document]
Comment (d4rkd3v1l) [reply Adyen/Components/Apple Pay/ApplePayComponent.swift:144]: Same here.

File: Adyen/Components/Apple Pay/ApplePayComponent.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (mohammedDehairy) [reply Adyen/Components/Apple Pay/ApplePayComponent.swift:101]: Thanks @d4rkd3v1l for your contribution, its highly appreciated.

The docs says:

`- Parameter completion: Completion handler being called on completion, for success as well as error/cancel.`

This is not true, this closure is not used in case payment is concluded either successful or failure, as in `stopLoading` the closure provided by the user during initialization will be overridden by the closure passed to `stopLoading`, and the one provided by the user will `only` be executed when user cancels the flow before calling `stopLoading`.

As for the naming, users will potentially think that `completion` closure is used when payment is completed, and I am afraid will use it on that assumption, but if its called `dismissHandler`, there is no mistake that it will only be executed in case of user canceled the Apple Pay flow.

[the remark this answers is not in the document]
Comment (d4rkd3v1l) [reply Adyen/Components/Apple Pay/ApplePayComponent.swift:101]: Fair enough... but what if you don't call `stopLoading`? Then there is no override, and the callback is being used in `paymentAuthorizationViewControllerDidFinish`.
For my understanding `stopLoading` is used, if I manually want to stop the process (which not really gonna happen for ApplePay)?! But I have to admit, that I'm not completely into this, right now.

Anyway, I don't know about any Adyen styleguide, conventions or whatever, and honestly I also really don't want to know about it. Furthermore I would appreciate if I don't have to know anything about the internal logic at all, and it would just work, which it unfortunately did not in this case. 

So basically I just want to use this (actually paid) service.
I like things being open sourced, but I don't fully get the business model behind having a paid service with open sourced SDKs?!

Maybe I should have just reported an issue, instead of fixing it myself.

[the remark this answers is not in the document]
Comment (mohammedDehairy) [reply Adyen/Components/Apple Pay/ApplePayComponent.swift:101]: calling's `stopLoading` is required after you send the Apple Pay token to your server and your server call the `/payments` Adyen endpoint as mentioned in the [docs](https://docs.adyen.com/payment-methods/apple-pay/ios-component#make-payment).

 We're sorry that you find the SDK not working as expected, we're working hard to provide the best possible experience for our merchants, and we provide the SDK as a major convenience for our merchants for smooth checkout experience.

Also we highly appreciate open source contributions, as we're big believers in open source culture, so thank you again for your contribution, and we'll continue to work hard to provide the best possible experience for you and all our merchants.

[the remark this answers is not in the document]
Comment (d4rkd3v1l) [reply Adyen/Components/Apple Pay/ApplePayComponent.swift:101]: Ok, somehow I completely missed that "stopLoading" requirement. I just wonder how ApplePay seems to work without it anyway. 😬🙈
However, I didn't want to offend you, sorry for that. 
So how do we proceed here, or what do you think is the (best) solution? Basically, as described in my initial message, I need to know, when the dialog get's dismissed, which at least was not the case, when one tabs cancel on the apple dialog. 
If you think my solution is fine, I indeed can just rename the stuff and try to find a more fitting/true description for the docs, otherwise feel free to propose another solution.

[the remark this answers is not in the document]
Comment (mohammedDehairy) [reply Adyen/Components/Apple Pay/ApplePayComponent.swift:101]: no problem at all, your solution is good, if you just rename the completion closure to `dismissHandler` or `cancelHandler`, and update the docs, then everything should be okay.

otherwise if you don't have time, its also completely okay if we fix this ourselves.

[the remark this answers is not in the document]
Comment (d4rkd3v1l) [reply Adyen/Components/Apple Pay/ApplePayComponent.swift]: I included your suggestions now, please let me know if there's anything else or if I missed s.th.

**Comment by descorp** [comment]
Congratulations @d4rkd3v1l
You have become the first official contributor to Public Adyen Repo! :D

Thank you for your time and your trust!
